:recycle: Architectural refactoring

yanang007 opened the pull request:
This PR aims to clarify the semantics and improve extensibility of current compensator's API.

Details
--------
1. Keeps original `build/apply` functions to provide semantic naming for parameters and its behaviors.

```python
def build(self, scalar, vector_x, vector_y, vector_z): ...
def apply(self, flux_x, flux_y, flux_z, op): ...
```

2. Implement sklearn-like `fit/transform` paradigm.

```python
def fit(self, X, y): ...
def transform(self, X, y): ...  # TollesLawsonCompensator / fit_transform
def predict(self, X, y): ...  # TollesLawson / fit_predict
def score(self, X, y): ...
```

3. Separate the noise prediction model (TollesLawson) from the compensation model (TollesLawsonCompensator).

```python
class TollesLawson: ...  # predicts the interference
class TollesLawsonCompensator(TollesLawson): ...  # transforms the data to de-interference
```

TODO
-------
1. Is it neccessary to introduce a new class like `AttitudePredictor` to seperate the process for predicting attitude (directional cosines)?

yanang007 committed:
:recycle: 仿照sklearn范式重构TollesLawsonCompensator

:recycle: 拆分TollesLawson预测模型和补偿模型
:heavy_plus_sign: 添加typing-extensions以兼容Python 3.11以前的版本中使用Self返回类型

dyuu7 committed:
:recycle: 以sklearn范式拆分磁三分量估计姿态特征功能

dyuu7 committed:
:recycle: 仿照sklearn.metrics拆分磁补偿性能评估相关功能

dyuu7 committed:
:recycle: 仿照sklearn参数校验方式调整磁总场补偿模块

dyuu7 committed:
:recycle: 调整metrics结构

dyuu7 committed:
:recycle: 细分utils功能

dyuu7 committed:
:recycle: 各式姿态估计器继承自基类以支持多态

dyuu7 committed:
:recycle: 定义磁补偿可能需要的数据种类，简化姿态特征估计方式

dyuu7 committed:
:recycle: 提取线性模型物理项，补偿器内部业务以多态方式运行，优化命名空间结构

dyuu7 committed:
:pencil: 接口与架构变化后的新tmi补偿示例

tangmqsoa commented on deinterf/compensation.py:
```diff
@@ -0,0 +1,280 @@
+    Parameters
+    ----------
+    with_band_pass_filter : bool, default=True
+        是否对输入数据进行带通滤波，滤波频率范围为[0.1, 0.5] Hz
```
这个频率该如何从实际数据计算，因为输入数据只是三个分量和总场值，不同的采样率下，该怎么设置这里的滤波频率参数？

dyuu7 committed:
:recycle: IoC方式重构底层数据模型

:sparkles: 实现IoC访问依赖日志
:recycle: 调整传感器数据类型和补偿项实现
:recycle: 根据新数据底层调整sklearn-like接口实现
:recycle: 将MagModulus重命名为MagIntensity以保持一致性
:memo: IOC数据形式下的新tmi补偿示例
:memo: 以拓展INS姿态为例展示用法
:see_no_evil: 避免追踪launch和env相关文件
:fire: 删除重复部分

Co-authored-by: yanang007 <yanang007@users.noreply.github.com>
